File: core/services/link.go
package services

import (
	"slimlink/core/entities"
	"slimlink/core/exceptions"
	"slimlink/core/ports"
	"slimlink/core/utils"
)

type linkService struct {
	linkRepo     ports.LinkRepo
	linkIDLength int
}

func NewLinkService(linkRepo ports.LinkRepo, linkIDLength int) ports.LinkService {
	return &linkService{linkRepo, linkIDLength}
}

func (s *linkService) CreateLink(url string) (*entities.Link, error) {
	if !utils.IsValidUrl(url) {
		return nil, exceptions.NewAppError[*exceptions.BadRequestError]("invalid URL")
	}
	var id string
	for attempt := 1; ; attempt++ {
		id = utils.GenerateBase62String(s.linkIDLength)
		_, err := s.linkRepo.GetByID(id)
		if err != nil {
			if exceptions.IsAppError[*exceptions.NotFoundError](err) {
				break
			}
			return nil, err
		}
		if attempt == 10 {
			return nil, exceptions.NewAppError[*exceptions.UnexpectedError]("maximum tries exceeded for link ID generation")
		}
	}
	link := &entities.Link{ID: id, Url: url}
	err := s.linkRepo.Add(link)
	if err != nil {
		return nil, err
	}
	return link, nil
}

func (svc *linkService) GetLinkByID(id string) (*entities.Link, error) {
	return svc.linkRepo.GetByID(id)
}

File: web/src/server/services/link.ts
import { getRedisKey, setRedisKey } from "server/data/redis";
import { BadRequestError, NotFoundError, UnexpectedError } from "server/exceptions";
import { generateBase62String } from "server/utils";
import { Link } from "types";
import { isValidUrl } from "utils";

const linkIDLength = (process.env.LINK_ID_LENGTH && parseInt(process.env.LINK_ID_LENGTH)) || 5;

export async function createLink(url: string): Promise<Link> {
	if (!isValidUrl(url)) {
		throw new BadRequestError("Invalid URL");
	}
	let id;
	for (let attempt = 1; ; attempt++) {
		id = generateBase62String(linkIDLength);
		try {
			// eslint-disable-next-line no-await-in-loop
			await getRedisKey(id);
		} catch (e) {
			if (e instanceof NotFoundError) {
				break;
			}
			throw e;
		}
		if (attempt === 10) {
			throw new UnexpectedError("Maximum tries exceeded for link ID generation");
		}
	}
	await setRedisKey(id, url);
	return { id, url: url };
}

export async function getLinkByID(id: string): Promise<Link> {
	const url = await getRedisKey(id);
	return { id, url };
}
